iceoryx_hoofs/source/posix_wrapper/unix_domain_socket.cpp

```diff
@@ -344,21 +364,20 @@ expected<std::string, IpcChannelError> UnixDomainSocket::timedReceive(const unit
     {
         return err(errnoToEnum(setsockoptCall.error().errnum));
     }
-    // NOLINTJUSTIFICATION needed for recvfrom
-    // NOLINTNEXTLINE(hicpp-avoid-c-arrays, cppcoreguidelines-avoid-c-arrays)
-    char message[MAX_MESSAGE_SIZE + 1];
 
-    auto recvCall = IOX_POSIX_CALL(iox_recvfrom)(m_sockfd, &message[0], MAX_MESSAGE_SIZE, 0, nullptr, nullptr)
+    auto recvCall = IOX_POSIX_CALL(iox_recvfrom)(m_sockfd, msg, msgSize, 0, nullptr, nullptr)
                         .failureReturnValue(ERROR_CODE)
                         .suppressErrorMessagesForErrnos(EAGAIN, EWOULDBLOCK)
                         .evaluate();
-    message[MAX_MESSAGE_SIZE] = 0;
-
     if (recvCall.has_error())
     {
+        // NOLINTNEXTLINE(cppcoreguidelines-pro-bounds-pointer-arithmetic)
+        msg[0] = 0;
         return err(errnoToEnum(recvCall.error().errnum));
     }
-    return ok<std::string>(&message[0]);
+    // NOLINTNEXTLINE(cppcoreguidelines-pro-bounds-pointer-arithmetic)
+    msg[static_cast<uint64_t>(recvCall->value) + NULL_TERMINATOR_SIZE] = 0;
```

Comment:
To avoid duplicated code, I add a boolean option
```
expected<void, PosixIpcChannelError> sendImpl(const char* buf, uint64_t bufSize, const units::Duration& timeout, const bool hasTerminator) const noexcept;

expected<uint64_t, PosixIpcChannelError> receiveImpl(char* buf, uint64_t maxBufSize, const units::Duration& timeout, const bool addTerminator) const noexcept;
```

Reply:
Here is my proposal:

```c++
enum class Termination {
NONE,
NULL_TERMINATOR, // STRING or STRING_TERMINATOR might also work
};

template <Typename Type>
expected<void, PosixIpcChannelError> sendImpl(const Type* msg, uint64_t msgSize, const units::Duration& timeout, const Termination terminator) const noexcept;

template <Typename Type>
expected<uint64_t, PosixIpcChannelError> receiveImpl(Type* msg, uint64_t maxBufSize, const units::Duration& timeout, const Termination terminator) const noexcept
```

The Template pointer is needed in the future if we want to send a vector of ```std::byte```, ```unit8_t``` etc.

If ```sendImpl``` is called with ```NULL_TERMINATOR```, the termination size (1) is added to ```msgSize``` to send also the terminator.

if ```receiveImpl``` is called with ``NULL_TERMINATOR```, the last buffer data arrived is checked has a terminator, otherwise, I return an error.

Reply:
This would work. With the additional template parameter there is a new option. If we would assume that a `char` type would always be null-terminated, we could omit the enum and just check for the type. This makes it less flexible though since it will not be possible to send a vector of chars anymore.

I think your proposal should work better. We also need to keep in mind to not make the code too complex just to prevent a bit of code duplication but up until now it is still okay.